// jet-validators/dist/basic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNullableStringArray = exports.isOptionalStringArray = exports.isStringArray = exports.isNullishString = exports.isNullableString = exports.isOptionalString = exports.isString = exports.isNishValidNumArr = exports.isNullableValidNumberArray = exports.isOptionalValidNumberArray = exports.isValidNumberArray = exports.isNullishValidNumber = exports.isNullableValidNumber = exports.isOptionalValidNumber = exports.isValidNumber = exports.isNullishBigIntArr = exports.isNullableBigIntArray = exports.isOptionalBigIntArray = exports.isBigIntArray = exports.isNullishBigInt = exports.isNullableBigInt = exports.isOptionalBigInt = exports.isBigInt = exports.isNullishNumberArray = exports.isNullableNumberArray = exports.isOptionalNumberArray = exports.isNumberArray = exports.isNullishNumber = exports.isNullableNumber = exports.isOptionalNumber = exports.isNumber = exports.isNullishValidBooleanArray = exports.isNullableValidBooleanArray = exports.isOptionalValidBooleanArray = exports.isValidBooleanArray = exports.isNullishValidBoolean = exports.isNullableValidBoolean = exports.isOptionalValidBoolean = exports.isValidBoolean = exports.isNullishBooleanArray = exports.isNullableBooleanArray = exports.isOptionalBooleanArray = exports.isBooleanArray = exports.isNullishBoolean = exports.isNullableBoolean = exports.isOptionalBoolean = exports.isBoolean = exports.isNullOrUndef = exports.isNull = exports.isUndef = void 0;
exports.isFunction = exports.isNullishRecordArray = exports.isNullableRecordArray = exports.isOptionalRecordArray = exports.isRecordArray = exports.isNullishRecord = exports.isNullableRecord = exports.isOptionalRecord = exports.isRecord = exports.isNullishObjectArray = exports.isNullableObjectArray = exports.isOptionalObjectArray = exports.isObjectArray = exports.isNullishObject = exports.isNullableObject = exports.isOptionalObject = exports.isObject = exports.isNullishValidDateArray = exports.isNullableValidDateArray = exports.isOptionalValidDateArray = exports.isValidDateArray = exports.isNullishValidDate = exports.isNullableValidDate = exports.isOptionalValidDate = exports.isValidDate = exports.isNullishDateArray = exports.isNullableDateArray = exports.isOptionalDateArray = exports.isDateArray = exports.isNullishDate = exports.isNullableDate = exports.isOptionalDate = exports.isDate = exports.isNullishSymbolArray = exports.isNullableSymbolArray = exports.isOptionalSymbolArray = exports.isSymbolArray = exports.isNullishSymbol = exports.isNullableSymbol = exports.isOptionalSymbol = exports.isSymbol = exports.isNullishNonEmptyStringArray = exports.isNullableNonEmptyStringArray = exports.isOptionalNonEmptyStringArray = exports.isNonEmptyStringArray = exports.isNullishNonEmptyString = exports.isNullableNonEmptyString = exports.isOptionalNonEmptyString = exports.isNonEmptyString = exports.isNullishStringArray = void 0;
exports.isNullishFunctionArray = exports.isNullableFunctionArray = exports.isOptionalFunctionArray = exports.isFunctionArray = exports.isNullishFunction = exports.isNullableFunction = exports.isOptionalFunction = void 0;
const common_1 = require("./common");
const utils_1 = require("../utils");
exports.isUndef = ((arg) => arg === undefined);
exports.isNull = ((arg) => arg === null);
exports.isNullOrUndef = (0, common_1.orNullable)(exports.isUndef);
exports.isBoolean = _checkType('boolean');
exports.isOptionalBoolean = (0, common_1.orOptional)(exports.isBoolean);
exports.isNullableBoolean = (0, common_1.orNullable)(exports.isBoolean);
exports.isNullishBoolean = (0, common_1.orNullable)(exports.isOptionalBoolean);
exports.isBooleanArray = _toArray(exports.isBoolean);
exports.isOptionalBooleanArray = (0, common_1.orOptional)(exports.isBooleanArray);
exports.isNullableBooleanArray = (0, common_1.orNullable)(exports.isBooleanArray);
exports.isNullishBooleanArray = (0, common_1.orNullable)(exports.isOptionalBooleanArray);
exports.isValidBoolean = _isValidBoolean;
exports.isOptionalValidBoolean = (0, common_1.orOptional)(exports.isValidBoolean);
exports.isNullableValidBoolean = (0, common_1.orNullable)(exports.isValidBoolean);
exports.isNullishValidBoolean = (0, common_1.orNullable)(exports.isOptionalValidBoolean);
exports.isValidBooleanArray = _toArray(exports.isValidBoolean);
exports.isOptionalValidBooleanArray = (0, common_1.orOptional)(exports.isValidBooleanArray);
exports.isNullableValidBooleanArray = (0, common_1.orNullable)(exports.isValidBooleanArray);
exports.isNullishValidBooleanArray = (0, common_1.orNullable)(exports.isOptionalValidBooleanArray);
exports.isNumber = _isNumber;
exports.isOptionalNumber = (0, common_1.orOptional)(exports.isNumber);
exports.isNullableNumber = (0, common_1.orNullable)(exports.isNumber);
exports.isNullishNumber = (0, common_1.orNullable)(exports.isOptionalNumber);
exports.isNumberArray = _toArray(exports.isNumber);
exports.isOptionalNumberArray = (0, common_1.orOptional)(exports.isNumberArray);
exports.isNullableNumberArray = (0, common_1.orNullable)(exports.isNumberArray);
exports.isNullishNumberArray = (0, common_1.orNullable)(exports.isOptionalNumberArray);
exports.isBigInt = _checkType('bigint');
exports.isOptionalBigInt = (0, common_1.orOptional)(exports.isBigInt);
exports.isNullableBigInt = (0, common_1.orNullable)(exports.isBigInt);
exports.isNullishBigInt = (0, common_1.orNullable)(exports.isOptionalBigInt);
exports.isBigIntArray = _toArray(exports.isBigInt);
exports.isOptionalBigIntArray = (0, common_1.orOptional)(exports.isBigIntArray);
exports.isNullableBigIntArray = (0, common_1.orNullable)(exports.isBigIntArray);
exports.isNullishBigIntArr = (0, common_1.orNullable)(exports.isOptionalBigIntArray);
exports.isValidNumber = _isValidNumber;
exports.isOptionalValidNumber = (0, common_1.orOptional)(exports.isValidNumber);
exports.isNullableValidNumber = (0, common_1.orNullable)(exports.isValidNumber);
exports.isNullishValidNumber = (0, common_1.orNullable)(exports.isOptionalValidNumber);
exports.isValidNumberArray = _toArray(exports.isValidNumber);
exports.isOptionalValidNumberArray = (0, common_1.orOptional)(exports.isValidNumberArray);
exports.isNullableValidNumberArray = (0, common_1.orNullable)(exports.isValidNumberArray);
exports.isNishValidNumArr = (0, common_1.orNullable)(exports.isOptionalValidNumberArray);
exports.isString = _checkType('string');
exports.isOptionalString = (0, common_1.orOptional)(exports.isString);
exports.isNullableString = (0, common_1.orNullable)(exports.isString);
exports.isNullishString = (0, common_1.orNullable)(exports.isOptionalString);
exports.isStringArray = _toArray(exports.isString);
exports.isOptionalStringArray = (0, common_1.orOptional)(exports.isStringArray);
exports.isNullableStringArray = (0, common_1.orNullable)(exports.isStringArray);
exports.isNullishStringArray = (0, common_1.orNullable)(exports.isOptionalStringArray);
exports.isNonEmptyString = _isNonEmptyString;
exports.isOptionalNonEmptyString = (0, common_1.orOptional)(exports.isNonEmptyString);
exports.isNullableNonEmptyString = (0, common_1.orNullable)(exports.isNonEmptyString);
exports.isNullishNonEmptyString = (0, common_1.orNullable)(exports.isOptionalNonEmptyString);
exports.isNonEmptyStringArray = _toArray(exports.isNonEmptyString);
exports.isOptionalNonEmptyStringArray = (0, common_1.orOptional)(exports.isNonEmptyStringArray);
exports.isNullableNonEmptyStringArray = (0, common_1.orNullable)(exports.isNonEmptyStringArray);
exports.isNullishNonEmptyStringArray = (0, common_1.orNullable)(exports.isOptionalNonEmptyStringArray);
exports.isSymbol = _checkType('symbol');
exports.isOptionalSymbol = (0, common_1.orOptional)(exports.isSymbol);
exports.isNullableSymbol = (0, common_1.orNullable)(exports.isSymbol);
exports.isNullishSymbol = (0, common_1.orNullable)(exports.isOptionalSymbol);
exports.isSymbolArray = _toArray(exports.isSymbol);
exports.isOptionalSymbolArray = (0, common_1.orOptional)(exports.isSymbolArray);
exports.isNullableSymbolArray = (0, common_1.orNullable)(exports.isSymbolArray);
exports.isNullishSymbolArray = (0, common_1.orNullable)(exports.isOptionalSymbolArray);
exports.isDate = _isDate;
exports.isOptionalDate = (0, common_1.orOptional)(exports.isDate);
exports.isNullableDate = (0, common_1.orNullable)(exports.isDate);
exports.isNullishDate = (0, common_1.orNullable)(exports.isOptionalDate);
exports.isDateArray = _toArray(exports.isDate);
exports.isOptionalDateArray = (0, common_1.orOptional)(exports.isDateArray);
exports.isNullableDateArray = (0, common_1.orNullable)(exports.isDateArray);
exports.isNullishDateArray = (0, common_1.orNullable)(exports.isOptionalDateArray);
exports.isValidDate = _isValidDate;
exports.isOptionalValidDate = (0, common_1.orOptional)(exports.isValidDate);
exports.isNullableValidDate = (0, common_1.orNullable)(exports.isValidDate);
exports.isNullishValidDate = (0, common_1.orNullable)(exports.isOptionalValidDate);
exports.isValidDateArray = _toArray(exports.isValidDate);
exports.isOptionalValidDateArray = (0, common_1.orOptional)(exports.isValidDateArray);
exports.isNullableValidDateArray = (0, common_1.orNullable)(exports.isValidDateArray);
exports.isNullishValidDateArray = (0, common_1.orNullable)(exports.isOptionalValidDateArray);
exports.isObject = _isObject;
exports.isOptionalObject = (0, common_1.orOptional)(exports.isObject);
exports.isNullableObject = (0, common_1.orNullable)(exports.isObject);
exports.isNullishObject = (0, common_1.orNullable)(exports.isOptionalObject);
exports.isObjectArray = _toArray(exports.isObject);
exports.isOptionalObjectArray = (0, common_1.orOptional)(exports.isObjectArray);
exports.isNullableObjectArray = (0, common_1.orNullable)(exports.isObjectArray);
exports.isNullishObjectArray = (0, common_1.orNullable)(exports.isOptionalObjectArray);
exports.isRecord = _isRecord;
exports.isOptionalRecord = (0, common_1.orOptional)(exports.isRecord);
exports.isNullableRecord = (0, common_1.orNullable)(exports.isRecord);
exports.isNullishRecord = (0, common_1.orNullable)(exports.isOptionalRecord);
exports.isRecordArray = _toArray(_isRecord);
exports.isOptionalRecordArray = (0, common_1.orOptional)(exports.isRecordArray);
exports.isNullableRecordArray = (0, common_1.orNullable)(exports.isRecordArray);
exports.isNullishRecordArray = (0, common_1.orNullable)(exports.isOptionalRecordArray);
exports.isFunction = _checkType('function');
exports.isOptionalFunction = (0, common_1.orOptional)(exports.isFunction);
exports.isNullableFunction = (0, common_1.orNullable)(exports.isFunction);
exports.isNullishFunction = (0, common_1.orNullable)(exports.isOptionalFunction);
exports.isFunctionArray = _toArray(exports.isFunction);
exports.isOptionalFunctionArray = (0, common_1.orOptional)(exports.isFunctionArray);
exports.isNullableFunctionArray = (0, common_1.orNullable)(exports.isFunctionArray);
exports.isNullishFunctionArray = (0, common_1.orNullable)(exports.isOptionalFunctionArray);
function _toArray(cb) {
    return (arg) => {
        return Array.isArray(arg) && !arg.some(item => !cb(item));
    };
}
function _isNumber(arg) {
    return typeof arg === 'number' && !isNaN(arg);
}
function _isValidNumber(arg) {
    const argf = Number(arg);
    return !isNaN(argf);
}
function _isObject(arg) {
    return typeof arg === 'object' && (arg !== null);
}
function _isRecord(arg) {
    return (0, exports.isObject)(arg) && !Array.isArray(arg);
}
function _checkType(type) {
    return (arg) => {
        return typeof arg === type;
    };
}
function _isValidBoolean(arg) {
    try {
        arg = (0, utils_1.parseBoolean)(arg);
        return (0, exports.isBoolean)(arg);
    }
    catch (err) {
        return false;
    }
}
function _isDate(arg) {
    return arg instanceof Date && !isNaN(arg.getTime());
}
function _isValidDate(arg) {
    if (!((0, exports.isString)(arg) || (0, exports.isNumber)(arg) || arg instanceof Date)) {
        return false;
    }
    const argf = new Date(arg);
    return !isNaN(argf.getTime());
}
function _isNonEmptyString(arg) {
    return ((0, exports.isString)(arg) && arg.length > 0);
}
